AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,210 @@
+      if (!error) {
+        UIImage *image = snapshot.image;
+        CGRect finalImageRect = CGRectMake(0, 0, image.size.width, image.size.height);
+        
+        UIGraphicsBeginImageContextWithOptions(image.size, YES, image.scale);
+        [image drawAtPoint:CGPointMake(0, 0)];
+        
+        if (_annotations.count > 0 ) {
+          // Get a standard annotation view pin. Future implementations should use a custom annotation image property.
+          MKAnnotationView *pin = [[MKPinAnnotationView alloc] initWithAnnotation:nil reuseIdentifier:@""];
+          UIImage *pinImage = pin.image;
+          for (id<MKAnnotation>annotation in _annotations)
+          {
+            CGPoint point = [snapshot pointForCoordinate:annotation.coordinate];
+            if (CGRectContainsPoint(finalImageRect, point))
+            {
+              CGPoint pinCenterOffset = pin.centerOffset;
+              point.x -= pin.bounds.size.width / 2.0;
+              point.y -= pin.bounds.size.height / 2.0;
+              point.x += pinCenterOffset.x;
+              point.y += pinCenterOffset.y;
+              [pinImage drawAtPoint:point];
+            }
+          }
+        }
+        
+        UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        self.image = finalImage;
+      }
+    }];
+  }
+}
+
+- (void)resetSnapshotter
+{
+  if (!_snapshotter.isLoading) {
+    _options.size = self.calculatedSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+#pragma mark - Action
+- (void)addLiveMap
+{
+  if (self.isNodeLoaded && !_mapNode) {
+    _mapNode = [[ASDisplayNode alloc]initWithViewBlock:^UIView *{
+      _mapView = [[MKMapView alloc]initWithFrame:CGRectMake(0.0f, 0.0f, self.calculatedSize.width, self.calculatedSize.height)];
+      _mapView.delegate = _mapDelegate;
+      [_mapView setRegion:_options.region];
+      [_mapView addAnnotations:_annotations];
+      return _mapView;
+    }];
+    [self addSubnode:_mapNode];
+    
+    if (CLLocationCoordinate2DIsValid(_centerCoordinateOfMap)) {
+      [_mapView setCenterCoordinate:_centerCoordinateOfMap];
+    }
+  }
+}
+
+- (void)removeLiveMap
+{
+  if (_mapNode) {
+    _centerCoordinateOfMap = _mapView.centerCoordinate;
+    [_mapNode removeFromSupernode];
+    _mapView = nil;
+    _mapNode = nil;
+  }
+  self.image = nil;
+}
+
+#pragma mark - Layout
+// Layout isn't usually needed in the box model, but since we are making use of MKMapView which is hidden in an ASDisplayNode this is preferred.
+- (void)layout
+{
+  [super layout];
+  if (_mapView) {
+    _mapView.frame = CGRectMake(0.0f, 0.0f, self.calculatedSize.width, self.calculatedSize.height);
+  }
+  else {
+    // If our bounds.size is different from our current snapshot size, then let's request a new image from MKMapSnapshotter.
+    if (!CGSizeEqualToSize(_options.size, self.bounds.size)) {
+      if (_needsMapReloadOnBoundsChange && self.image) {
```

Comment:
should we remove the check for self.image here?  What if I ask the map to reload on bounds change, and then I put it onscreen (and it starts loading) - then change its size before it finishes loading?